Assets/CameraController.cs

```diff
@@ -0,0 +1,151 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class CameraController : MonoBehaviour
+{
+    [SerializeField]
+    private Transform cameraTransform;
+
+    [SerializeField]
+    private float normalSpeed;
+    [SerializeField]
+    private float fastSpeed;
+    [SerializeField]
+    private float movementSpeed;
+    [SerializeField]
+    private float movementTime;
+    [SerializeField]
+    private float rotationAmount;
+    [SerializeField]
+    private Vector3 zoomAmount;
+
+    private Vector3 newPosition;
+    private Quaternion newRotation;
+    private Vector3 newZoom;
+
+    private Vector3 dragStartPosition;
+    private Vector3 dragCurrentPosition;
+    private Vector3 rotateStartPosition;
+    private Vector3 rotateCurrentPosition;
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        newPosition = transform.position;
+        newRotation = transform.rotation;
+        newZoom = cameraTransform.localPosition;
+    }
+
+    // Update is called once per frame
+    void Update()
+    {
+        HandleMouseInput();
+        HandleMovementInput();
+    }
+
+    void HandleMouseInput()
+    {
+        if (Input.mouseScrollDelta.y != 0)
+        {
+            newZoom += Input.mouseScrollDelta.y * zoomAmount;
+        }
+
+        if(Input.GetMouseButtonDown(0))
+        {
+            Plane plane = new Plane(Vector3.up, Vector3.zero);
+
+            Ray ray = Camera.main.ScreenPointToRay(Input.mousePosition);
+        
+            float entry;
+
+            if(plane.Raycast(ray, out entry))
+            {
+                dragStartPosition = ray.GetPoint(entry);
+            }
+        }
+
+        if(Input.GetMouseButton(0))
+        {
+            Plane plane = new Plane(Vector3.up, Vector3.zero);
+
+            Ray ray = Camera.main.ScreenPointToRay(Input.mousePosition);
+        
+            float entry;
+
+            if(plane.Raycast(ray, out entry))
+            {
+                dragCurrentPosition = ray.GetPoint(entry);
+
+                newPosition = transform.position + dragStartPosition - dragCurrentPosition;
+            }
+        }
+
+        if(Input.GetMouseButtonDown(2))
+        {
+            rotateStartPosition = Input.mousePosition;
+        }
+        
+        if(Input.GetMouseButton(2))
+        {
+            rotateCurrentPosition = Input.mousePosition;
+
+            Vector3 difference = rotateStartPosition - rotateCurrentPosition;
+
+            rotateStartPosition = rotateCurrentPosition;
+
+            newRotation *= Quaternion.Euler(Vector3.up * (-difference.x / 5f));
+        }
+    }
+
+    void HandleMovementInput()
+    {
+        if(Input.GetKey(KeyCode.LeftShift))
```

Comment:
Use ternary operator.
Also add Assert which checks if fastSpeed is greater than normalSpeed

Reply:
It should be done as validation in movement settings when you change fastSpeed, not when you try to move faster while playng.

Reply:
Agree. There is no mechanism for unity to do it in inspector. It could be added in Awake method.